src/SIMDMath/readme.md

```diff
@@ -59,22 +59,52 @@ julia> @btime test_simd(x) setup=(x=rand()*2)
 ```julia
-const P4 =  ntuple(n -> rand()*(-1)^n / n, 4)
-const P8 =  ntuple(n -> rand()*(-1)^n / n, 8)
-const P16 =  ntuple(n -> rand()*(-1)^n / n, 16)
-const P32 =  ntuple(n -> rand()*(-1)^n / n, 32)
-const P64 =  ntuple(n -> rand()*(-1)^n / n, 64)
-
-@btime evalpoly(x, P4) setup=(x=rand())
-@btime evalpoly(x, P8) setup=(x=rand())
-@btime evalpoly(x, P16) setup=(x=rand())
-@btime evalpoly(x, P32) setup=(x=rand())
-@btime evalpoly(x, P64) setup=(x=rand())
+let
+    horner_times = []
+    horner2_times = []
+    horner4_times = []
+    horner8_times = []
+
+    for N in [4, 8, 12, 16, 24, 48, 96, 198]
+        poly = ntuple(n -> rand()*(-1)^n / n, N)
+        poly_packed2 = pack_horner2(poly)
+        poly_packed4 = pack_horner4(poly)
+        poly_packed8 = pack_horner8(poly)
+
+        t1 = @benchmark evalpoly(x, $poly) setup=(x=rand())
+        t2 = @benchmark horner2(x, $poly_packed2) setup=(x=rand())
+        t3 = @benchmark horner4(x, $poly_packed4) setup=(x=rand())
+        t4 = @benchmark horner8(x, $poly_packed8) setup=(x=rand())
+
+        push!(horner_times,  round(minimum(t1).time, digits=3))
+        push!(horner2_times,  round(minimum(t2).time, digits=3))
+        push!(horner4_times,  round(minimum(t3).time, digits=3))
+        push!(horner8_times,  round(minimum(t4).time, digits=3))
+
+    end
+    @show horner_times
+    @show horner2_times
+    @show horner4_times
+    @show horner8_times
+end
 ```
 
 As mentioned, Horner's scheme requires sequential multiply-add instructions that can't be performed in parallel. One way (another way is Estrin's method which we won't discuss) to improve this structure is to break the polynomial down into even and odd polynomials (a second order Horner's scheme) or into larger powers of `x^4` or `x^8` (a fourth and eighth order Horner's scheme) which allow for computing many different polynomials of similar length simultaneously. In some regard, we are just rearranging the coefficients and using the same method as we did in the first case with some additional arithmetic at the end to add all the different polynomials together.
 
 The last fact is important because we are actually increasing the total amount of arithmetic operations needed but increasing by a large amount the number of operations that can happen in parallel. The increased operations make the advantages of this approach less straightforward than the first case which is always superior. The second and perhaps most important point is that floating point arithmetic is not associative so these approaches will give slightly different results as we are adding and multiplying in slightly differnet order.
 
+The above benchmarks are as follows.
+```julia
+Julia Version 1.8.2
+Platform Info:
+  OS: macOS (arm64-apple-darwin21.3.0)
+  CPU: 8 × Apple M1
+
+horner_times  = [2.125, 2.125, 2.416, 2.416, 4.000, 12.345, 40.573, 126.353]
```

Comment:
can you normalize these timings?

Reply:
Yes - good point. I also added a different CPU. Very interesting as it will actually do better than the expected 2x or 4x at high degrees (I'm assuming due to the better memory layout).

For small degrees it is just noisy though so hard to really compare. I'll need a better benchmark I think for this as we are trying to distinguish like one or two multiplies. Though it is cool to see even for really small degrees this matches the regular `evalpoly`.